sdk/storage/azure-storage-quickquery/src/main/java/com/azure/storage/quickquery/implementation/util/FluxInputStream.java

```diff
@@ -0,0 +1,220 @@
+
+    private ByteArrayInputStream buffer;
+
+    private boolean subscribed;
+    private boolean fluxComplete;
+    private boolean waitingForData;
+
+    private final Lock lock;
+    private final Condition dataAvailable;
+
+    private IOException lastError;
+
+    /**
+     * Creates a new FluxInputStream
+     *
+     * @param data The data to subscribe to and read from.
+     * @param logger {@link ClientLogger}
+     */
+    public FluxInputStream(Flux<ByteBuffer> data, ClientLogger logger) {
+        this.subscribed = false;
+        this.fluxComplete = false;
+        this.waitingForData = false;
+        this.data = data;
+        this.logger = logger;
+        this.lock = new ReentrantLock();
+        this.dataAvailable = lock.newCondition();
+    }
+
+    @Override
+    public int read() throws IOException {
+        return read(new byte[1], 0, 1);
+    }
+
+    @Override
+    public int read(byte[] b, int off, int len) throws IOException {
+        validateParameters(b, off, len);
+
+        /* If len is 0, then no bytes are read and 0 is returned. */
+        if (len == 0) {
+            return 0;
+        }
+        /* Attempt to read at least one byte. If no byte is available because the stream is at end of file,
+           the value -1 is returned; otherwise, at least one byte is read and stored into b. */
+
+        /* Not subscribed? subscribe and block for data */
+        if (!subscribed) {
+            blockForData();
+        }
+        /* Now, we have subscribed. */
+        /* At this point, buffer should not be null. If it is, that indicates either an error or completion event
+           was emitted by the Flux. */
+        if (this.buffer == null) { // Only executed on first subscription.
+            if (this.lastError != null) {
+                throw lastError;
+            }
+            if (this.fluxComplete) {
+                return -1;
+            }
+            this.buffer = new ByteArrayInputStream(new byte[0]);
+        }
+
+        /* Now we are guaranteed that buffer is SOMETHING. */
+        /* No data is available in the buffer.  */
+        if (this.buffer.available() == 0) {
+            /* If the flux completed, there is no more data available to be read from the stream. Return -1. */
+            if (this.fluxComplete) {
+                return -1;
+            }
+            /* If we are not waiting for another request to come in, request more data. */
+            if (!waitingForData) {
+                /* Block current thread until data is available. */
+                blockForData();
+            }
+        }
+
+        /* Data available in buffer, read the buffer. */
+        if (this.buffer.available() > 0) {
+            return this.buffer.read(b, off, len);
+        }
+
+        /* If the flux completed, there is no more data available to be read from the stream. Return -1. */
+        if (this.fluxComplete) {
+            return -1;
+        }
+
+        return 0;
```

Comment:
Under what conditions could we end up here?

Reply:
I think its an error - I'll throw an ioException but I'm not sure what the message should be - like this indicates that the flux just abruptly stopped? but I'm not sure if that's possible

Reply:
Again, I think IllegalStateException might be more descriptive. It doesn't have the same expectations as IOException for readable messages, so putting something vague and unhelpful like "Reached end of method. Should never happen," is more acceptable.